### ISSUE TITLE
panic: inconsistent list element types

### ISSUE THREAD

**Issue by Poil** [hashicorp/terraform#27275]
Hi,

I probably did something wrong because my element types are inconsistent, but should not crash

### Terraform Version

```
Terraform v0.14.2
+ provider registry.terraform.io/hashicorp/azurerm v2.40.0
+ provider registry.terraform.io/hashicorp/external v2.0.0
+ provider registry.terraform.io/hashicorp/random v3.0.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
+ provider registry.terraform.io/hashicorp/vault v2.16.0
+ provider registry.terraform.io/terraform-providers/mysql v1.9.0
```

### Crash Output

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"backends":cty.DynamicPseudoType, "health_probe_name":cty.String, "name":cty.String}) then cty.Object(map[string]cty.Type{"backends":cty.Tuple([]cty.Type{cty.Object(map[string]cty.Type{"address":cty.DynamicPseudoType, "host_header":cty.String})}), "health_probe_name":cty.String, "name":cty.String}))

goroutine 3021 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc002dbe380, 0x2, 0x2, 0xc001a11b20, 0x1, 0x1, 0x1)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.0/cty/value_init.go:166 +0x445
github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x2c51060, 0xc001a82da0, 0x2328540, 0xc001a82dc0, 0x0, 0x0, 0x0, 0xc000077800, 0x10, 0x10, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.0/cty/convert/conversion_collection.go:327 +0x794
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2c51060, 0xc001a82da0, 0x2328540, 0xc001a82dc0, 0x0, 0x0, 0x0, 0xc004621530, 0xc001a11b10, 0x2c50f60, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.0/cty/convert/conversion.go:46 +0x433
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x2c51060, 0xc001a82da0, 0x2328540, 0xc001a82dc0, 0xc001a11b10, 0x0, 0x0, 0x0, 0xc000b3c8f0, 0x10000c00293c200)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.0/cty/convert/conversion.go:188 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x2c51060, 0xc001a82da0, 0x2328540, 0xc001a82dc0, 0x2c50f60, 0xc0009fe8c0, 0xc001a82da0, 0x2328540, 0xc001a82dc0, 0x0, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.7.0/cty/convert/public.go:51 +0x1aa
github.com/hashicorp/terraform/terraform.(*nodeModuleVariable).EvalModuleCallArgument(0xc000b5f2c0, 0x2c8c480, 0xc0013b64e0, 0xc000115101, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/node_module_variable.go:238 +0x265
github.com/hashicorp/terraform/terraform.(*nodeModuleVariable).Execute(0xc000b5f2c0, 0x2c8c480, 0xc0013b64e0, 0xc001bce004, 0x24a0780, 0x25f4d80)
	/home/circleci/project/project/terraform/node_module_variable.go:157 +0x7f
github.com/hashicorp/terraform/terraform.(*ContextGraphWalker).Execute(0xc0018081a0, 0x2c8c480, 0xc0013b64e0, 0x7f279b0a50a8, 0xc000b5f2c0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph_walk_context.go:127 +0xbc
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x25f4d80, 0xc000b5f2c0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/graph.go:59 +0x962
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc000b5f320, 0x25f4d80, 0xc000b5f2c0, 0xc001c3be40)
	/home/circleci/project/project/dag/walk.go:387 +0x375
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/home/circleci/project/project/dag/walk.go:309 +0x1246



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.log" file that was created may contain 
sensitive information that must be redacted before it is safe to share 
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`

**Comment by Poil** [comment]
I also reproduce with TF 0.13.5

**Comment by alisdair** [comment]
Hi @Poil, thanks for reporting this. Terraform shouldn't crash like this, so we'd like to fix this bug.

To do so, we need a small but complete configuration which will reproduce this bug. Can you provide that for us in a reply to this issue, please?

**Comment by Poil** [comment]
I will try to extract the crashing part this week

**Comment by Poil** [comment]
The problem is in this block

```
  backend_pools = [
    {
      name              = local.varnish_backend_pool_name
      health_probe_name = local.varnish_health_probe_name
      backends = [for vm in module.varnish-vm :
        {
          host_header = ""
          address     = vm.vm_public_ip_address
        }
      ]
    },
    {
      name              = local.bo_backend_pool_name
      health_probe_name = local.bo_health_probe_name
      backends = [
        { 
          host_header = ""
          address     = module.lb-bo.pubip_ip_address
        }
      ]
    }
  ]
```

If I remove the first or the second part of the `backends` block it does not crash.

Best regards,

**Comment by alisdair** [comment]
Hi @Poil, thanks for the additional information. Unfortunately that isn't enough detail for us to reproduce the issue.

Can you produce a complete but small configuration which exhibits the same issue? That will let us see what's happening and hopefully fix the bug. Thanks!

**Comment by dthvt** [comment]
@alisdair reproduction example here: https://gist.github.com/dthvt/ce36da543abf22ad2ec2db8e7f9bcbaf

This appears to be some sort of interaction w/ module input typing and a list that has some elements with sub-elements composed by [for ...] statements.

**Comment by alisdair** [comment]
Thanks very much, @dthvt! I've confirmed the issue and slightly reduced your reproduction case:

main.tf
```hcl
resource "random_pet" "pets" {
  count = 2
}

module "foo" {
  source  = "./foo"
  configs = local.configs
}

locals {
  configs = [
    {
      names = ["alice", "bob"],
    },
    {
      names = [for pet in random_pet.pets : pet.id],
    }
  ]
}
```

foo/main.tf:
```hcl
variable "configs" {
  type = list(any)
}
```

This crashes with `terraform plan`, partial trace here:

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"names":cty.Tuple([]cty.Type{cty.String, cty.String})}) then cty.Object(map[string]cty.Type{"names":cty.DynamicPseudoType}))

goroutine 76 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc00014b9c0, 0x2, 0x2, 0xc000a02f30, 0x1, 0x1, 0x1)
	/Users/alisdair/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/value_init.go:166 +0x5a8
github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x38647e0, 0xc0002ca6a0, 0x2f38800, 0xc0002ca6c0, 0x0, 0x0, 0x0, 0x38646a0, 0xda988e0, 0x10, ...)
	/Users/alisdair/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/convert/conversion_collection.go:327 +0x794
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x38647e0, 0xc0002ca6a0, 0x2f38800, 0xc0002ca6c0, 0x0, 0x0, 0x0, 0xc000bfa180, 0xc000a02f20, 0x38646e0, ...)
	/Users/alisdair/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/convert/conversion.go:46 +0x433
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x38647e0, 0xc0002ca6a0, 0x2f38800, 0xc0002ca6c0, 0xc000a02f20, 0x495bfa0, 0x0, 0x0, 0xc000367ad0, 0x10000c00010c900)
	/Users/alisdair/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/convert/conversion.go:188 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x38647e0, 0xc0002ca6a0, 0x2f38800, 0xc0002ca6c0, 0x38646e0, 0xc000590e20, 0xc0002ca6a0, 0x2f38800, 0xc0002ca6c0, 0x0, ...)
	/Users/alisdair/go/pkg/mod/github.com/zclconf/go-cty@v1.7.1/cty/convert/public.go:51 +0x1b9
github.com/hashicorp/terraform/terraform.(*nodeModuleVariable).EvalModuleCallArgument(0xc000233920, 0x389ff80, 0xc0004d5790, 0xc000266001, 0x0, 0x2, 0x28)
	/Users/alisdair/code/terraform/terraform/node_module_variable.go:238 +0x265
```

**Comment by alisdair** [comment]
The issue is (as shown in the stack trace) in the upstream cty library. [I've proposed a fix over there](https://github.com/zclconf/go-cty/pull/85), but it will require review (and possibly a rethink) before we can bring the change back to Terraform.

**Comment by alisdair** [comment]
I've been revisiting this recently, and while I don't have a usable fix for the bug, I thought it might be useful to note a workaround. All instances of this crash that I can find are resolved by declaring a full type for the input variable.

Considering the reduced failure case:

```hcl
resource "random_pet" "pets" {
  count = 2
}

module "foo" {
  source  = "./foo"
  configs = local.configs
}

locals {
  configs = [
    {
      names = ["alice", "bob"],
    },
    {
      names = [for pet in random_pet.pets : pet.id],
    }
  ]
}
```

With this module definition, Terraform panics:

```hcl
variable "configs" {
  type = list(any)
}
```

If we instead specify the full type of the variable, Terraform is able to resolve the types correctly:

```hcl
variable "configs" {
  type = list(object({ names = list(string) }))
}
```

There are also other advantages to fully specifying the input variable type—earlier errors when the input value is invalid, for example.

**Comment by wenslayer** [comment]
I ran into this same panic with 0.14.10 while trying to use the [Experimental Optional Object Type Attributes](https://www.terraform.io/docs/language/expressions/type-constraints.html#experimental-optional-object-type-attributes) with this repro case:
```
$ terraform -version
Terraform v0.14.10

Your version of Terraform is out of date! The latest version
is 0.15.0. You can update by downloading from https://www.terraform.io/downloads.html

$ cat main.tf
terraform {
  experiments = [module_variable_optional_attrs]
  required_version = "~> 0.14"
}

variable "domains_map" {
  description = "list of domains and optional properties"

  type = list(
    object({
      name    = string
      comment = optional(string)
      dns_records = optional(list(
        object({
          name    = string
          records = list(string)
          ttl     = optional(string)
          type    = string
        })))}))
}

output "done" {
  value = "made it here"
}

$ cat terraform.tfvars
domains_map = [
    {
      name = "foo1.com"
    },
    {
      name    = "foo2.com"
      comment = "this is foo2"
    },
    {
      name    = "foo2.com"
      comment = "Foo3 comment"
      dns_records = [
        {
          name    = "@"
          type    = "TXT"
          records = ["v=spf1 -all"]
        }
      ]
    },
  ]

$ terraform plan
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"comment":cty.String, "dns_records":cty.List(cty.ObjectWithOptionalAttrs(map[string]cty.Type{"name":cty.String, "records":cty.List(cty.String), "ttl":cty.String, "type":cty.String}, []string{"ttl"})), "name":cty.String}) then cty.Object(map[string]cty.Type{"comment":cty.String, "dns_records":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "records":cty.List(cty.String), "ttl":cty.String, "type":cty.String})), "name":cty.String}))

goroutine 13 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc000279440, 0x3, 0x3, 0xc00081b7c0, 0x1, 0x1, 0x1)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/value_init.go:166 +0x5a8
github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x38e8420, 0xc000727a40, 0x2f89be0, 0xc000727a60, 0x0, 0x0, 0x0, 0x38e83a0, 0x5299390, 0x10, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/convert/conversion_collection.go:327 +0x794
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x38e8420, 0xc000727a40, 0x2f89be0, 0xc000727a60, 0x0, 0x0, 0x0, 0xc0002ec060, 0xc00081b7b0, 0x38e8320, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/convert/conversion.go:46 +0x433
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x38e8420, 0xc000727a40, 0x2f89be0, 0xc000727a60, 0xc00081b7b0, 0xc0001db5f0, 0xc00020f818, 0x0, 0xc0002b0008, 0x1010425)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/convert/conversion.go:188 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x38e8420, 0xc000727a40, 0x2f89be0, 0xc000727a60, 0x38e8320, 0xc00081b560, 0x0, 0xc000182d00, 0xc000314e68, 0x10de3d7, ...)
	/go/pkg/mod/github.com/zclconf/go-cty@v1.8.0/cty/convert/public.go:51 +0x1b9
github.com/hashicorp/terraform/terraform.checkInputVariables(0xc0001db5f0, 0xc0001dbce0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/terraform/variables.go:255 +0x193
github.com/hashicorp/terraform/terraform.NewContext(0xc0003159d0, 0x1, 0x1, 0xc000315900, 0x1)
	/home/circleci/project/project/terraform/context.go:279 +0x11f3
github.com/hashicorp/terraform/backend/local.(*Local).contextDirect(0xc0001d7c70, 0xc0001b8c40, 0xc00020f930, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0001db8c0, 0xc0007270c0, ...)
	/home/circleci/project/project/backend/local/backend_local.go:178 +0x2ff
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc0001d7c70, 0xc0001b8c40, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/circleci/project/project/backend/local/backend_local.go:111 +0xf18
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc0001d7c70, 0x38e62a0, 0xc000238980, 0x38e62a0, 0xc0002389c0, 0xc0001b8c40, 0xc000238940)
	/home/circleci/project/project/backend/local/backend_plan.go:70 +0x2cb
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc00081b1f0, 0xc00081b200, 0xc00081b210, 0xc0001d7c70, 0xc00081b1e0, 0x38e62a0, 0xc000238980, 0x38e62a0, 0xc0002389c0, 0xc0001b8c40, ...)
	/home/circleci/project/project/backend/local/backend.go:340 +0xe8
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
	/home/circleci/project/project/backend/local/backend.go:334 +0x36d
```

However, it works with 0.15.0:
```
$ terraform -version
Terraform v0.15.0
on darwin_amd64

$ terraform plan

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:

Terraform will perform the following actions:

Plan: 0 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + done = "made it here"
╷
│ Warning: Experimental feature "module_variable_optional_attrs" is active
│
│   on main.tf line 2, in terraform:
│    2:   experiments = [module_variable_optional_attrs]
│
│ Experimental features are subject to breaking changes in future minor or patch releases, based on feedback.
│
│ If you have feedback on the design of this feature, please open a GitHub issue to discuss it.
╵
```

**Comment by mildwonkey** [comment]
This particular panic is fixed in v0.15, and there is a known workaround for earlier releases, so I am going to close this issue. Thanks everyone for the information!
